### PR TITLE
MSN (Masked Siamese Networks) for ViT

### LINKED ISSUE #18758: Add MSN checkpoints to ViT
### Feature request

It's been a while Meta released the [MSN](https://github.com/facebookresearch/msn) (Masked Siamese Networks) checkpoints. MSN is known to perform pretty well in few-shot settings. Few-shot regime is important for real-world applications. They have only released the encoder weights.

I have worked on a [Colab Notebook](https://colab.research.google.com/gist/sayakpaul/51556ce57facea02041cf0c0f1c3fed3/scratchpad.ipynb) that converts the released MSN checkpoints into HF format and performs assertions to ensure the converted weights can be loaded into HF's ViT classes.

This issue proposes adding the MSN checkpoints to be compatible with HF's ViT classes. 

### Motivation

MSN is particularly useful in few-shot learning regimes.

### Your contribution

Can contribute to the checkpoints and other related utilities.

@NielsRogge could you assign this issue to me?

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Adds the [MSN](https://arxiv.org/abs/2204.07141) checkpoints for ViT. MSN shines in the few-shot regimes which would benefit real-world use cases. Later we could add a pre-training script so that people can actually perform pre-training with MSN with their own datasets. 

Closes #18758

## Who can review?

@sgugger @NielsRogge @amyeroberts 

## TODO

- [x] Add documentation
- [x] Add rest of the files for repo consistency 
- [ ] Host MSN weights on the Facebook org on HF Hub (@NielsRogge ?)
- [ ] Change the checkpoint paths wherever needed

### CODE REVIEW

**Comment by sayakpaul** [comment]
@NielsRogge, after studying the [pretraining script of MSN](https://github.com/facebookresearch/msn/blob/main/src/msn_train.py) thoroughly I am still unsure of how to put together a `ViTMSNForPretraining` similar to `ViTMAEForPreTraining`. There are multiple moving pieces that I think are best off residing inside a standalone pretraining script:

* A target encoder [updated with EMA](https://github.com/facebookresearch/msn/blob/main/src/msn_train.py#L373).
* [Learnable prototypes](https://github.com/facebookresearch/msn/blob/main/src/msn_train.py#L217) that are needed to compute the final MSN loss. 
* [Target sharpening](https://github.com/facebookresearch/msn/blob/main/src/msn_train.py#L347) amongst other things.  

Both the EMA and sharpening components operate with their own schedules. 

Given this, I think it's best to resort to a separate pre-training script and use this model for feature extraction and fine-tuning. 

There's an [ongoing discussion](https://github.com/facebookresearch/msn/issues/7) around releasing the weights of the linear classification layers and fine-tuned models. So when that's available, we could directly support those via `ViTMSNForImageClassification`. Regardless, I am happy to add a `ViTMSNForImageClassification` for easy access. 

What do you think?

**Comment by sgugger** [comment]
Thanks for your PR! It would be great to have the `ViTMSNForImageClassification` even if there are no released weights for image classification, so users can already fine-tune the main checkpoint if they want.

For pretraining, if multiple new pieces are needed, maybe it could go in a research project at first, where you can add more modules?

**Comment by sayakpaul** [comment]
> For pretraining, if multiple new pieces are needed, maybe it could go in a research project at first, where you can add more modules?

Sounds good to me. 

> Thanks for your PR! It would be great to have the ViTMSNForImageClassification even if there are no released weights for image classification, so users can already fine-tune the main checkpoint if they want.

Sure, I will continue the work from here on then. Thank you!

File: src/transformers/models/vit_msn/configuration_vit_msn.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+VIT_MSN_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "sayakpaul/vit-msn-base": "https://huggingface.co/sayakpaul/vit-msn-base/resolve/main/config.json",
```
Comment (sayakpaul) [inline src/transformers/models/vit_msn/configuration_vit_msn.py:25]: To be changed.

File: src/transformers/models/vit_msn/modeling_vit_msn.py [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+from .configuration_vit_msn import ViTMSNConfig
+
+
+logger = logging.get_logger(__name__)
+
+
+_CONFIG_FOR_DOC = "ViTMSNConfig"
+_CHECKPOINT_FOR_DOC = "sayakpaul/vit-msn-small"
+VIT_MSN_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "sayakpaul/vit-msn-small",
```
Comment (sayakpaul) [inline src/transformers/models/vit_msn/modeling_vit_msn.py:41]: To be changed.

File: tests/models/vit_msn/test_modeling_vit_msn.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
+    return image
+
+
+@require_torch
+@require_vision
+class ViTMSNModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        return ViTFeatureExtractor.from_pretrained("sayakpaul/vit-msn-small") if is_vision_available() else None
```
Comment (sayakpaul) [inline tests/models/vit_msn/test_modeling_vit_msn.py:218]: To be changed.

File: tests/models/vit_msn/test_modeling_vit_msn.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    @slow
+    def test_inference_image_classification_head(self):
+        torch.manual_seed(2)
+        model = ViTMSNForImageClassification.from_pretrained("sayakpaul/vit-msn-small").to(torch_device)
```
Comment (sayakpaul) [inline tests/models/vit_msn/test_modeling_vit_msn.py:223]: To be changed.

File: tests/models/vit_msn/test_modeling_vit_msn.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+@require_torch
+@require_vision
+class ViTMSNModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        return ViTFeatureExtractor.from_pretrained("sayakpaul/vit-msn-small") if is_vision_available() else None
+
+    @slow
+    def test_inference_image_classification_head(self):
+        torch.manual_seed(2)
```
Comment (sayakpaul) [inline tests/models/vit_msn/test_modeling_vit_msn.py:222]: This is here so as to ensure the classification head params are always initialized using the same init. Since we don't have classification head params from the authors yet, this is needed to ensure the tests are passing.

**Comment by sayakpaul** [comment]
@sgugger @NielsRogge @amyeroberts ready for review.

**Comment by sayakpaul** [comment]
@sgugger @NielsRogge @amyeroberts a friendly nudge on the PR.

File: src/transformers/models/vit_msn/configuration_vit_msn.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            The dropout ratio for the attention probabilities.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-06):
+            The epsilon used by the layer normalization layers.
+        image_size (`int`, *optional*, defaults to `224`):
+            The size (resolution) of each image.
+        patch_size (`int`, *optional*, defaults to `16`):
+            The size (resolution) of each patch.
+        num_channels (`int`, *optional*, defaults to `3`):
```
Comment (sgugger) [inline src/transformers/models/vit_msn/configuration_vit_msn.py:65]:
```suggestion
        image_size (`int`, *optional*, defaults to 224):
            The size (resolution) of each image.
        patch_size (`int`, *optional*, defaults to 16):
            The size (resolution) of each patch.
        num_channels (`int`, *optional*, defaults to 3):
```
Numbers should not be in code blocks.

File: src/transformers/models/vit_msn/modeling_vit_msn.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+        if num_patches == num_positions and height == width:
+            return self.position_embeddings
+        class_pos_embed = self.position_embeddings[:, 0]
+        patch_pos_embed = self.position_embeddings[:, 1:]
+        dim = embeddings.shape[-1]
+        h0 = height // self.config.patch_size
+        w0 = width // self.config.patch_size
+        # we add a small number to avoid floating point error in the interpolation
+        # see discussion at https://github.com/facebookresearch/dino/issues/8
+        h0, w0 = h0 + 0.1, w0 + 0.1
```
Comment (sgugger) [inline src/transformers/models/vit_msn/modeling_vit_msn.py:82]: Please use better variable names here

File: src/transformers/models/vit_msn/modeling_vit_msn.py [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+        patch_pos_embed = patch_pos_embed.reshape(1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim)
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed,
+            scale_factor=(h0 / math.sqrt(num_positions), w0 / math.sqrt(num_positions)),
+            mode="bicubic",
+            align_corners=False,
+        )
+        assert int(h0) == patch_pos_embed.shape[-2] and int(w0) == patch_pos_embed.shape[-1]
```
Comment (sgugger) [inline src/transformers/models/vit_msn/modeling_vit_msn.py:91]: No new assert in the code base. Please use a test and raise an appropriate error.

**Review by sgugger — APPROVED** [review]
A couple more comments, then the weights should be transferred to the right org before we merge this PR.

**Comment by sayakpaul** [comment]
@sgugger addressed your comments. After the weights are transferred to the right org, I will open a PR there adding README.

**Review by amyeroberts — APPROVED** [review]
Very nice PR ❤️  All LGTM. 

Thanks for your work adding this model 💪

File: src/transformers/models/vit_msn/modeling_vit_msn.py [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/vit-msn-small")
+        >>> with torch.no_grad():
+        ...     model = ViTMSNModel.from_pretrained("facebook/vit-msn-small")
+
+        >>> inputs = feature_extractor(images=image, return_tensors="pt")
+        >>> outputs = model(**inputs)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_msn/modeling_vit_msn.py:548]: No. Nice catch!

File: src/transformers/models/vit_msn/modeling_vit_msn.py [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+            return head_outputs + encoder_outputs[1:]
+
+        return BaseModelOutput(
+            last_hidden_state=sequence_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
+
+
+# Caution: We don't have the weights for the classification head yet. This class
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_msn/modeling_vit_msn.py:591]: Thanks! I am contacting the model author Mido to check if they can release the classification params before we merge. Will keep y'all posted.

**Comment by ydshieh** [comment]
Hi @sayakpaul . First, thank you for this PR 🤗 .

The doctest for this model is currently failing, as 
https://github.com/huggingface/transformers/blob/7e84723fe4e9a232e5e27dc38aed373c0c7ab94a/src/transformers/models/vit_msn/modeling_vit_msn.py#L646
this outputs the predicted label, but there is no expected value provided.

The [config](https://huggingface.co/facebook/vit-msn-small/blob/main/config.json) has `LABEL_0` ... `LABEL_999` in `id2label`, but I feel it should be the actual labels for the COCO dataset. 

Could you take a look for this config, as well as the missing expected outputs for the doctest? Thank you!

Here is the failing doctest job:

https://github.com/huggingface/transformers/actions/runs/3109562462/jobs/5039877349

**Comment by sayakpaul** [comment]
> The config has LABEL_0 ... LABEL_999 in id2label, but I feel it should be the actual labels for the COCO dataset. 

The model was trained on ImageNet-1k. 

I will add the expected outputs. Thanks for flagging it.
